benchexec/resources.py

```diff
@@ -210,118 +218,738 @@ def _get_cpu_cores_per_run0(
+
+    # compare num of units & runs per unit vs num_of_threads
+    if len(hierarchy_levels[chosen_level]) * runs_per_unit < num_of_threads:
+        if not isTest:
+            sys.exit(
+                f"Cannot assign required number of threads."
+                f"Please reduce the number of threads to {len(hierarchy_levels[chosen_level]) * runs_per_unit}."
+            )
+        else:
+            is_feasible = False
+
+    # calculate if sub_units have to be split to accommodate the runs_per_unit
+    sub_units_per_run = calculate_sub_units_per_run(
+        coreLimit_rounded_up, hierarchy_levels, chosen_level
+    )
+    # number of nodes at subunit-Level / sub_units_per_run
+    if len(hierarchy_levels[chosen_level - 1]) / sub_units_per_run < num_of_threads:
+        if not isTest:
+            sys.exit(
+                f"Cannot split memory regions between runs. "
+                f"Please reduce the number of threads to {math.floor(len(hierarchy_levels[chosen_level-1]) / sub_units_per_run)}."
             )
+        else:
+            is_feasible = False
+
+    return is_feasible
+
+
+def calculate_chosen_level(
+    hierarchy_levels: List[HierarchyLevel], coreLimit_rounded_up: int
+) -> int:
+    """Calculates the hierarchy level necessary so that number of cores at the chosen_level is at least
+    as big as the cores necessary for one thread"""
+    chosen_level = 1
+    # move up in hierarchy as long as the number of cores at the current level is smaller than the coreLimit
+    # if the number of cores at the current level is as big as the coreLimit: exit loop
+    while (
+        chosen_level < len(hierarchy_levels) - 1
+        and len(next(iter(hierarchy_levels[chosen_level].values())))
+        < coreLimit_rounded_up
+    ):
+        chosen_level = chosen_level + 1
+    return chosen_level
+
+
+def calculate_coreLimit_rounded_up(
+    siblings_of_core: HierarchyLevel, coreLimit: int
+) -> int:
+    """coreLimit_rounded_up (int): recalculate # cores for each run accounting for HT"""
+    core_size = len(next(iter(siblings_of_core.values())))
+    # Take value from hierarchy_levels instead from siblings_of_core
+    coreLimit_rounded_up = int(math.ceil(coreLimit / core_size) * core_size)
+    assert coreLimit <= coreLimit_rounded_up < (coreLimit + core_size)
+    return coreLimit_rounded_up
+
+
+def calculate_sub_units_per_run(
+    coreLimit_rounded_up: int,
+    hierarchy_levels: List[HierarchyLevel],
+    chosen_level: int,
+) -> int:
+    """calculate how many sub_units have to be used to accommodate the runs_per_unit"""
+    sub_units_per_run = math.ceil(
+        coreLimit_rounded_up / len(hierarchy_levels[chosen_level - 1][0])
+    )
+    return sub_units_per_run
+
+
+def check_and_add_meta_level(
+    hierarchy_levels: List[HierarchyLevel], allCpus: Dict[int, VirtualCore]
+) -> None:
+    """
+    Adds a meta_level to hierarchy_levels to iterate through all cores (if necessary)
```

Comment:
What is a `meta_level` and what purpose does it have? This term never occurs anywhere else.

Reply:
Adds a meta_level or root_level which includes all cores to hierarchy_levels (if necessary).
This is necessary to iterate through all cores if the highest hierarchy level consists of more than one unit.
Also adds the identifier for the new level to the memory region of all cores in allCpus

Reply:
Improved by 2f3f9640ee20cce5a124e2903accd71f51b1238b.